src/Microsoft.OData.Core/Json/DefaultStreamBasedJsonWriterFactory.cs

```diff
@@ -0,0 +1,35 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="DefaultStreamBasedJsonWriterFactory.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP3_1_OR_GREATER
+using System;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.OData.Json
+{
+
+    /// <summary>
+    /// Default factory for JSON writer that writes directly to a stream
+    /// rather than a TextWriter.
+    /// </summary>
+    [CLSCompliant(false)]
+    public sealed class DefaultStreamBasedJsonWriterFactory : IStreamBasedJsonWriterFactory
+    {
+        public IJsonWriter CreateJsonWriter(Stream stream, bool isIeee754Compatible, Encoding encoding)
+        {
+            // TODO: only supports Utf8-encoding for now. Support for other encodings
+            // to be added later.
+            if (encoding.EncodingName != Encoding.UTF8.EncodingName)
+            {
+                return null;

```

Comment:
Is it better to throw an exception?

Reply:
I did also think of throwing an exception, and this feels like a hack. The reason I didn't throw an exception is to make it easier to fall back to the default `JsonWriter` without having to resort to a try-catch in the `ODataJsonLightOutputContext.CreateJsonWriter` method. Since technically other encodings are supported.

Another option I considered is to return a `JsonWriter` directly.

What do you think?

Reply:
@habbes Maybe state in the `summary` or `remark` for the method doc to explain that the writer currently handles `utf-8` encoding only and a null is returned if any other encoding is passed

Reply:
@habbes Could you perhaps apply some null checks for `stream` and `encoding` just in case someone passed the two arguments as null to `CreateJsonWriter` method? Especially for `stream` argument you should throw an exception

Reply:
@corranrogue9 this class is the `DefaultStreamBasedJsonWriterFactory`, it's the one that returns null when it meets an unsupported encoding.

I think what you mean is to use the current `DefaultJsonWriterFactory`. The problem is that it doesn't take a stream as input, it takes a `TextWriter`. This is the reason why I add a new `StreamBased` interface to avoid breaking changes. So if I call the current default factory, I would need to wrap the stream in `TextWriter` and pass to the `DefaultJsonWriterFactory.CreateJsonWriter`. Since this TextWriter would be created in this local scope, then the created `JsonWriter` would be the only one holding a reference to it. For this reason `JsonWriter` would need to dispose of the `TextWriter`. Currently, the `JsonWriter` does not assume ownership of the `TextWriter`. So I would need to add a flag to the `JsonWriter` constructor, and to the `DefaultJsonWriterFactory.CreateJsonWriter` to instruct it whether or not it should dispose the `TextWriter` so that it can support both the existing usage where it's not the owner, and this new usage where it has to be the owner otherwise we'll end up not disposing `TextWriter`.

So there's some complexity introduced in going that route.

Bear in mind that this approach of returning `null` is only temporary. As mentioned in the comments, I will add support for other encodings. I just didn't want to do all the work that needs to be done in a single PR because it will end up being too big and complex.

Reply:
Let me first address this ownership question. A `JsonWriter` can have a constructor overload that takes in a `Stream` and instantiates its own `StreamWriter`. It would then own the `StreamWriter`, and there's absolutely no confusion, and no compatibility issues because `JsonWriter` is already `IDisposable`. It's best to completely avoid these `shouldDispose` type flags, and we certainly shouldn't surface in a factory interface an intermediate object. 

Regarding the `null` or `Exception` issue, this is the scenario as I understand it: a customer either does or doesn't configure a `IStreamBasedJsonWriterFactory` and/or a `IJsonWriterFactory`. 
1. Let's take the case where they don't configure either. In this case, we want to create a `JsonWriter`.
2. In the case where they configure only a `IStreamBasedJsonWriterFactory`, then their factory has three options
    1. It correctly supports the request, and the customer gives us a `IJsonWriter`
    2. It doesn't support the request, and returns `null` based on the contract laid out by the factory interface. 
    3. It doesn't support the request, and throws an exception because the factory didn't account for that branch

    In the case of i, we just use their `IJsonWriter`. To me, ii and iii are undefined, but should probably be handled the same way. I could see if the customer's factory can't handle the request, we should give them a default implementation, or if the factory can't handle the request, we immediately through because there is a bug and the customer needs to address that. There can also be other options, but ultimately I see them being handle by ODL in the same way.
3. In the case where they configure only a `IJsonWriterFactory`, then their factory has the same three options as 2
4. In the case where they configure both, I'm a little unsure, but I'm going to go out on a limb and say this is actually an error. This should not happen. We should probably validate against it, and even so, in the code that leverages the factory, we shouldn't attempt to account for this. If there is a `IStreamBasedJsonWriterFactory`, then we *assume* there is no `IJsonWriterFactory`. A customer can do anything with a `IStreamBasedJsonWriterFactory` that they could have done with a `IJsonWriterFactory`, and if they have a current implementation of `IJsonWriterFactory`, then can compose it into their `IStreamBasedJsonWriterFactory` implementation.

Because of all of this, I would expect code (ODL's or otherwise) that leverages these factories to look something like (and I will assume the behavior for 2ii and 2iii is to give a default implementation):
```
IJsonWriter Retrieve(Stream stream, Encoding encoding, params)
{
  var streamFactory = container.GetService<IStreamBasedJsonWriterFactory>();
  if (streamFactory == null)
  {
    this.textWriter = new StreamWriter(outputStream, encoding);
    var textWriterFactory = container.GetService<IJsonWriterFactory>();
    if (textWriterFactory == null)
    {
        return new JsonWriter(this.textWriter, params);
    }
    else
    {
      IJsonWriter jsonWriter;
      try
      {
        jsonWriter = textWriterFactory.CreateJsonWriter(this.textWriter, params);
      }
      catch
      {
        return new JsonWriter(this.textWriter, params);
      }

      if (jsonWriter == null)
      {
        return new JsonWriter(this.textWriter, params);
      }

      return jsonWriter;
    }
  }

  IJsonWriter jsonWriter;
  try
  {
    jsonWriter = streamFactory.CreateJsonWriter(stream, encoding, params);
  }
  catch
  {
    return new JsonWriter(this.textWriter, params);
  }

  if (jsonWriter == null)
  {
    return new JsonWriter(this.textWriter, params);
  }

  return jsonWriter;
}
```
However, if we change the contract of `IStreamBasedJsonWriterFactory` to throw if something is unsupported, the code would look like:
```
IJsonWriter Retrieve(Stream stream, Encoding encoding, params)
{
  var streamFactory = container.GetService<IStreamBasedJsonWriterFactory>();
  if (streamFactory == null)
  {
    this.textWriter = new StreamWriter(outputStream, encoding);
    var textWriterFactory = container.GetService<IJsonWriterFactory>();
    if (textWriterFactory == null)
    {
        return new JsonWriter(this.textWriter, params);
    }
    else
    {
      try
      {
        return textWriterFactory.CreateJsonWriter(this.textWriter, params);
      }
      catch
      {
        return new JsonWriter(this.textWriter, params);
      }
    }
  }

  try
  {
    return streamFactory.CreateJsonWriter(stream, encoding, params);
  }
  catch
  {
    return new JsonWriter(this.textWriter, params);
  }
}
```
Normally, I would not suggest that we use `Exception`s for control flow logic. In this case, I still agree, but we've actually removed the need for the `Exception` to represent control flow logic: if the contract of `IStreamBasedJsonWriterFactory` is such that it shouldn't return `null`, then our default implementation would need to be updated to only throw in catastrophic cases. If we make that change, then there is no real reason that the `ODataJsonLightOutputContext` needs to provide the default value for `null` or an `Exception`: the default factory contains the logic for the default cases. Any customer can also leverage the default factory if they need, so we should also only be expected exceptions from them in exceptional cases, rather than as control flow logic to indicate to use the default value.

Something that I'm noticing is that we are currently ignoring the `Exception` cases from a factory. We probably shouldn't ignore those cases. If we do choose to keep the `null` branching, then I to use the "Try{Method}" convention, so something like: `bool TryCreateJsonWriter(Stream stream, Encoding encoding, ..., out IJsonWriter jsonWriter`. I think this better fits the scenario we have. I will also say, though, that the convention seems to be such that the "Try{Method}" method shouldn't throw, so that would still be a bit strange.

Reply:
@corranrogue9 regarding the ownership issue, yes the `JsonWriter` can have a constructor overload that accepts a stream and instantiates its own TextWriter. However, the public facing `IJsonWriterFactory` and its default implementation accepts a `TextWriter`.

As for the factory returning null. That's a stop-gap temporary solution. I have already added support for other encodings on a separate branch: https://github.com/habbes/odata.net/tree/feat/2424-odatautf8jsonwriter-encodings. It uses a transcoding stream that converts from utf-8 to the target encoding internally, it has no dependency on the existing `JsonWriter` or `IJsonWriterFactory`.

I also don't think it's a good idea to rely on Exception for control-flow in this case.

If the customer provides their own `IStreamBasedJsonWriterFactory` that throws an exception, I don't think we should handle that exception or fallback to a default writer, I think we should let that exception bubble up and let the customer handle it.

With the new branch, the default stream-based writer factory is no longer null, and is not expected to throw an exception. If it does throw an exception, this is probably due to something else that went wrong and I don't think we should try to fallback to a defaut writer in this case either. Let the caller handle the exception as they see it fit.

If the customer provides an `IStreamBasedJsonWriterFactory` that returns null, we can treat that the same way as the existing `IJsonWriterFactory`. Currently ODL just has a `Debug.Assert(jsonWriter != null, ...)` check. Changing that to an exception would be a breaking change.

As for validating the scenario where a customer registers both an `IStreamBasedJsonWriterFactory` and `IJsonWriterFactory`, do you suggesting throwing an exception? If so, I can do that in the new branch.

Reply:
> As for validating the scenario where a customer registers both an IStreamBasedJsonWriterFactory and IJsonWriterFactory, do you suggesting throwing an exception? If so, I can do that in the new branch.

If we don't handle any of the other cases, then I don't see why we would add any code for this case. 

> If the customer provides their own IStreamBasedJsonWriterFactory that throws an exception, I don't think we should handle that exception or fallback to a default writer, I think we should let that exception bubble up and let the customer handle it.

If we are expecting that they can write an exception that we shouldn't handle, then we should probably also expect that they return `null` in a situation that we shouldn't handle. Never expecting `null` and relying on an exception takes away any ambiguity. A `TryCreate` method could also help here, but then the resulting `out IJsonWriter` could *also* accidentally be `null`. Whatever we do, if the implementer returns `null` when they don't mean to, we will default and they will never know. 

This is "contrary" to what I was saying before because now I'm saying we shouldn't provide a default value, but my point is really that we should do the same thing for both cases. Either, we should use a default if they return `null` or throw an exception, or we should throw if they return `null` or throw an exception. We cannot differentiate between these as extreme error cases compared to control flow logic.

Reply:
In the existing implementation, ODL doesn't do anything (other than a debug assert) if the user-provided `IJsonWriterFactory` returns null. So likely the code will break with a `NullReferenceException` at some later point. I think changing this behaviour will result in a breaking change.

What I can do is replicate similar behaviour for the new `IStreamBasedJsonWriterFactory`. But I'd still suggest to do that in the new branch since that one already implements support for other encodings.

Reply:
Alternatively I could merge that other branch into this one so it all goes in one PR, if people are okay with reviewing a larger PR at a go.

Reply:
@corranrogue9  I add support for different encodings to this PR

I also added validation to prevent the user from injecting both an `IJsonWriterFactory` and `IStreamBasedJsonWriterFactory`, but I rolled back that change after realizing that a default `IJsonWriterFactory` is injected in the container by default when you call `AddDefaultODataServices()`. And services that use `ODataMessageWriter`, including AspNetCoreOData, rely on this method.

So here's the current behaviour:
if there's an `IStreamBasedJsonWriterFactory` registered in the container, then it will use that one to create the writer. If it's not injected , then it will fallback to the current implementation.

It does not expect `IStreamBasedJsonWriterFactory`. If their factory happens to return null, this is not something we check for (which is consistent with the current behaviour), there's a `Debug.Assert` that checks this, but we don't throw an exception. So this will eventually result in a `NullReferenceException`, which is consistent with the current behaviour.

Reply:
Thanks Clement, I'm satisfied. @xuzhg @gathogojr can we mark this thread resolved?